pkg/workflow/controllers/parallel_node_reconciler.go

```diff
@@ -0,0 +1,245 @@
+	err := it.kubeClient.Get(ctx, request.NamespacedName, &node)
+	if err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// only resolve parallel nodes
+	if node.Spec.Type != v1alpha1.TypeParallel {
+		return reconcile.Result{}, nil
+	}
+
+	it.logger.V(4).Info("resolve parallel node", "node", request)
+
+	// make effects, create/remove children nodes
+	err = it.syncChildrenNodes(ctx, node)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// update status
+	updateError := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		nodeNeedUpdate := v1alpha1.WorkflowNode{}
+		err := it.kubeClient.Get(ctx, request.NamespacedName, &nodeNeedUpdate)
+		if err != nil {
+			return err
+		}
+
+		activeChildren, finishedChildren, err := it.fetchChildrenNodes(ctx, nodeNeedUpdate)
+		if err != nil {
+			return err
+		}
+
+		nodeNeedUpdate.Status.FinishedChildren = nil
+		for _, finishedChild := range finishedChildren {
+			nodeNeedUpdate.Status.FinishedChildren = append(nodeNeedUpdate.Status.FinishedChildren,
+				corev1.LocalObjectReference{
+					Name: finishedChild.Name,
+				})
+		}
+
+		nodeNeedUpdate.Status.ActiveChildren = nil
+		for _, activeChild := range activeChildren {
+			nodeNeedUpdate.Status.ActiveChildren = append(nodeNeedUpdate.Status.ActiveChildren,
+				corev1.LocalObjectReference{
+					Name: activeChild.Name,
+				})
+		}
+
+		// TODO: also check the consistent between spec in task and the spec in child node
+		if len(finishedChildren) == len(nodeNeedUpdate.Spec.Tasks) {
+			SetCondition(&nodeNeedUpdate.Status, v1alpha1.WorkflowNodeCondition{
+				Type:   v1alpha1.ConditionAccomplished,
+				Status: corev1.ConditionTrue,
+				Reason: "",
+			})
+		} else {
+			SetCondition(&nodeNeedUpdate.Status, v1alpha1.WorkflowNodeCondition{
+				Type:   v1alpha1.ConditionAccomplished,
+				Status: corev1.ConditionFalse,
+				Reason: "",
+			})
+		}
+
+		return it.kubeClient.Status().Update(ctx, &nodeNeedUpdate)
+	})
+
+	if updateError != nil {
+		it.logger.Error(err, "failed to update the status of node", "node", request)
+		return reconcile.Result{}, updateError
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (it *ParallelNodeReconciler) syncChildrenNodes(ctx context.Context, node v1alpha1.WorkflowNode) error {
+
+	// empty parallel node
+	if len(node.Spec.Tasks) == 0 {
+		it.logger.V(4).Info("empty parallel node, NOOP",
+			"node", fmt.Sprintf("%s/%s", node.Namespace, node.Name),
+		)
+		return nil
+	}
+
+	activeChildrenNodes, finishedChildrenNodes, err := it.fetchChildrenNodes(ctx, node)
+	if err != nil {
+		return err
+	}
+	existsChildrenNodes := append(activeChildrenNodes, finishedChildrenNodes...)
+
+	var taskNamesOfNodes []string
+	for _, childNode := range existsChildrenNodes {
+		taskNamesOfNodes = append(taskNamesOfNodes, getTaskNameFromGeneratedName(childNode.GetName()))
+	}
+
+	var tasksToStartup []string
+	if len(relativeComplementSet(taskNamesOfNodes, node.Spec.Tasks)) > 0 {
```

Comment:
What if I only add one task? Would the entire node be restarted?

Reply:
Yes. I think it might be better if all of the tasks will be spawned at the same time and run **parallelly**.

Reply:
But it seems that if I only add one task, `relativeComplementSet(taskNamesOfNodes, node.Spec.Tasks)` would be empty. Thus it would not be restarted.

Reply:
I update that logic, PTAL